core/src/main/scala/org/apache/spark/scheduler/LiveListenerBus.scala

```diff
@@ -39,98 +40,114 @@ import org.apache.spark.util.Utils
  * is stopped when `stop()` is called, and it will drop further events after stopping.
  */
-private[spark] class LiveListenerBus(conf: SparkConf) extends SparkListenerBus {
-
-  self =>
+private[spark] class LiveListenerBus(conf: SparkConf)
+  extends WithListenerBus[SparkListenerInterface, SparkListenerEvent] with Logging{
 
   import LiveListenerBus._
 
   private var sparkContext: SparkContext = _
+  private var metricsSystem: MetricsSystem = _
 
-  // Cap the capacity of the event queue so we get an explicit error (rather than
-  // an OOM exception) if it's perpetually being added to more quickly than it's being drained.
-  private val eventQueue =
-    new LinkedBlockingQueue[SparkListenerEvent](conf.get(LISTENER_BUS_EVENT_QUEUE_CAPACITY))
+  private val defaultListenerPool = GroupOfListener("default")
+  private val defaultListenerQueue = BusQueue(
+    conf.get(LISTENER_BUS_EVENT_QUEUE_CAPACITY),
+    defaultListenerPool,
+    BusQueue.ALL_MESSAGES)
 
-  private[spark] val metrics = new LiveListenerBusMetrics(conf, eventQueue)
+  @volatile private var otherListenerQueues = Seq.empty[BusQueue]
 
-  // Indicate if `start()` is called
-  private val started = new AtomicBoolean(false)
-  // Indicate if `stop()` is called
+  // start, stop and add/remove listener should be mutually exclusive
+  private val startStopAddRemoveLock = new ReentrantLock()
+  // Will be set modified in a synchronized function
+  @volatile private var started = false
   private val stopped = new AtomicBoolean(false)
 
-  /** A counter for dropped events. It will be reset every time we log it. */
-  private val droppedEventsCounter = new AtomicLong(0L)
-
-  /** When `droppedEventsCounter` was logged last time in milliseconds. */
-  @volatile private var lastReportTimestamp = 0L
-
-  // Indicate if we are processing some event
-  // Guarded by `self`
-  private var processingEvent = false
-
-  private val logDroppedEvent = new AtomicBoolean(false)
+   /**
+    * if isolatedIfPossible is true, add the listener to an isolated pool.
+    * Otherwise add it to the default pool.
+    * This method is thread-safe and can be called in any thread.
+    */
+  final override def addListener(listener: SparkListenerInterface,
+                                 isolatedIfPossible: Boolean): Unit = {
+    startStopAddRemoveLock.lock()
+    Try{
+      if (isolatedIfPossible) {
+         addQueue(BusQueue(
+          conf.get(LISTENER_BUS_EVENT_QUEUE_CAPACITY),
+          listener,
+          BusQueue.ALL_MESSAGES))
+      } else {
+        defaultListenerPool.addListener(listener)
+      }
+    }
+    startStopAddRemoveLock.unlock()
+  }
 
-  // A counter that represents the number of events produced and consumed in the queue
-  private val eventLock = new Semaphore(0)
+   /**
+    * Add a generic listener to an isolated pool.
+    */
+  def addProcessor(processor: SparkListenerEvent => Unit,
```

Comment:
with addProcessor, you do not have to provide a SparkListenerInterface (So with a method per message type), but just a generic function which handle  SparkListenerEvent  (the super type of each event type). So when you do a generic processing (see EventLoggingListener for example)  it is very convenient, and cherry on the top you avoid the horrible and costly dispatch function , which in this case (generic processing) is a burden